src/lib/blkalloc/append_blk_allocator.cpp

```diff
@@ -102,86 +80,68 @@ BlkAllocStatus AppendBlkAllocator::alloc(blk_count_t nblks, const blk_alloc_hint
     }
 
     // Push 1 blk to the vector which has all the requested nblks;
-    out_bid = BlkId{m_last_append_offset, nblks, m_chunk_id};
-
-    auto cur_cp = hs()->cp_mgr().cp_guard();
-    m_last_append_offset += nblks;
-    m_freeable_nblks -= nblks;
-
-    // it is guaranteed that dirty buffer always contains updates of current_cp or next_cp, it will
-    // never get dirty buffer from across updates;
-    set_dirty_offset(cur_cp->id() % MAX_CP_COUNT);
+    out_bid = BlkId{m_last_append_offset.fetch_add(nblks), nblks, m_chunk_id};
 
     COUNTER_INCREMENT(m_metrics, num_alloc, 1);
 
     return BlkAllocStatus::SUCCESS;
 }
 
-BlkAllocStatus AppendBlkAllocator::alloc_on_disk(BlkId const&) { return BlkAllocStatus::SUCCESS; }
-
-BlkAllocStatus AppendBlkAllocator::mark_blk_allocated(BlkId const&) { return BlkAllocStatus::SUCCESS; }
-
-void AppendBlkAllocator::free_on_disk(BlkId const&) {}
+BlkAllocStatus AppendBlkAllocator::alloc_on_disk(BlkId const&) {
+    m_is_dirty.store(true);
+    return BlkAllocStatus::SUCCESS;
+}
 
-bool AppendBlkAllocator::is_blk_alloced_on_disk(BlkId const&, bool) const { return false; }
+BlkAllocStatus AppendBlkAllocator::alloc_on_cache(BlkId const& blkid) {
+    auto new_offset = blkid.blk_num() + blkid.blk_count();
+    auto cur_offset = m_last_append_offset.load();
+    do {
+        if (cur_offset >= new_offset) { break; } // Already allocated
+    } while (!m_last_append_offset.compare_exchange_weak(cur_offset, new_offset));
+    return BlkAllocStatus::SUCCESS;
+}
 
-//
-// cp_flush doesn't need CPGuard as it is triggered by CPMgr which already handles the reference check;
-//
-// cp_flush should not block alloc/free;
-//
 void AppendBlkAllocator::cp_flush(CP* cp) {
-    const auto idx = cp->id() % MAX_CP_COUNT;
     // check if current cp's context has dirty buffer already
-    if (m_dirty_sb[idx].is_dirty) {
-        m_sb->last_append_offset = m_dirty_sb[idx].last_append_offset;
-        m_sb->freeable_nblks = m_dirty_sb[idx].freeable_nblks;
+    if (m_is_dirty.exchange(false)) {
+        m_sb->last_append_offset = m_last_append_offset.load();
```

Comment:
It is okay to do so by removing the m_dirty_sb, but it will take the latest m_last_append_offset that belongs to next CP, which is fine, that we need to do some check on alloc_on_cache during recovery (pls see my other comment).